### DISCUSSION TITLE
Preserving changes made to database models

### DISCUSSION THREAD

**Discussion by sc15zs** [communitiesuk/funding-service-requests-for-comments#15] [General]
Owner: zoltan.szabo@communities.gov.uk
Created on: 2024-12-12
Status: Draft
Finalised on: TBC

# Overview
Auditing and maintaining a trail of changes to database models is an important practice for ensuring accountability and traceability. Tracking fields and keeping a record of changes to our models is something we haven’t been doing but will likely need sooner or later.
At present, editing model fields is limited to developers. However, as (or if?) we expand our Flask-Admin functionality, non-developers may also need to make changes to models in the future. This makes it important to address preserving and tracking changes.

## What is the Current State?
Right now, we don’t have any system in place for tracking changes to models. If someone edits a model through the admin view or elsewhere, the change is applied directly to the database without leaving a trace. There’s no way to see what changed, when it changed, or who made the change.
Well, this is not entirely correct. We rely on log files to preserve details about issues such as key errors, upload errors, validation errors, exceptions, and similar problems. In a few places, we also use logging to record changes, but this approach isn’t consistent or comprehensive. Logs are helpful, but they’re not a substitute for a structured change-tracking system.
At the moment, non-developers don’t have access to edit model fields, but this will likely change. Without implementing change tracking, we may eventually face challenges in determining when specific changes were made and by whom. Either we won’t be able to find this information at all, or we’ll have to sift through log files in search of clues.

## Why Should We Change?
Without tracking changes, we’re blind to important details about updates in the system:
1. Auditing Challenges: If something goes wrong, we’ll have no way to trace changes back to their source.
2. Accountability: Knowing who made changes and why is essential for transparency.
3. Future-Proofing: Implementing change tracking ensures we have a history of changes. The earlier we have some kind of a tracking system in place, the better. 

## How Should We Address This?
We need a system that tracks changes to models efficiently and reliably. Here are a few approaches to consider:
1. Generic Table for Changes
    * A single table that records all changes to all models we want to track.
    * Fields might include a Foreign key to the model, field name, old value, new value, timestamp, and user ID.
    * Pros: Centralised, everything is stored in one table, easy implementation .
    * Cons: Could become unwieldy with frequent updates across many models, and the table size may grow large over time, potentially impacting performance.
2. Per-Model History Tables
    * Create a history model for each model we want to track.
    * Similarly to the previous solution, fields might include a Foreign key to the model, field name, old value, new value, timestamp, and user ID.
    * Pros: Organised, each model has its own specific history table, easy implementation.
    * Cons: ?
3. Third-Party Libraries
    * Use existing solutions like SQLAlchemy plugins or other libraries designed for change tracking.
    * Pros: probably a well-tested solution.
    * Cons: May not be as flexible for our specific needs.

Each of these options has trade-offs, so we’ll need to discuss and decide which one best fits our use case. 
Other solutions may certainly exist, please add these in your comments.

## What Alternatives Are We Discarding?
Doing nothing or relying on recording changes in log files only. Without a proper system, we’ll have no way to trace changes.

## Who Will Be Affected?
Nobody would be directly affected by this implementation. It would work in the background, recording changes without impacting current workflows. There wouldn’t be any changes for developers or non-developers.

## Who Will Benefit?
The Funding Service as a whole will benefit by gaining:
* Improved auditing and compliance reporting.
* Greater accountability for anyone making changes.

## What Are the Key Risks to Manage or Mitigate?
1. Performance Overhead: Tracking changes might slow down database operations?
    * Mitigation: In case of a custom solution, creating one record for each update is not a real issue. I _think_ it’s also the case with third-party solutions. 
2. Storage Growth: Tracking table(s) might grow too large over time.
    * Mitigation: Archive old records periodically to keep the active database manageable.
3. Implementation Errors: If the system isn’t designed properly, it might fail to capture all changes accurately.
    * Mitigation: Testing (ugh…) and validation during implementation.
4. Future Maintenance: Either custom solutions or third-party libraries may introduce additional technical debt.
    * Mitigation: Use modular designs, keeping the solution as simple as possible and build something which will be suitable in the long term.

**Comment by samuelhwilliams** [comment]
Hey @sc15zs - thanks for raising this, really good to see!

In the "how should we address this" section you've listed a few options. Is this RFC about "should we track changes?" and looking for consensus around a "yes/no" decision, or is this about a specific _way_ of tracking changes - ie deciding which of your three options we should go for?

**Reply by sc15zs** [reply]
@samuelhwilliams 
Hi Sam, and thanks for your comment.

The list in the "how should we address this" section is definitely non-exhaustive, these were the options I could think of. 
I assumed that we want to have some kind of change tracking in place, mainly because in the admin views anyone with edit permission can change some fields. 

If you think that it would be better to decide first whether or not we want to implement something to address the issue, we can change it to do so, and then decide about what way do do it if the decision is positive.

**Reply by samuelhwilliams** [reply]
I think it's fairly reasonable to assume that we _do_ want this tracking, so maybe this RFC should be about agreeing the specific implementation we want to use?

Do you think you're in a position to suggest which option we might want to take, or do we still need to explore that further?

**Reply by sc15zs** [reply]
I don't feel that I'm the right person to suggest which option we should use, I'm too biased towards the custom solution (without even trying what could be done e.g. with a SqlAlchemy library). 
If someone has a suggestion which pre-built solution would be suitable for us, I think it would be useful to try that and compare with a custom solution. 
What do you think?

**Reply by samuelhwilliams** [reply]
I think it's ok to be biased - we all are :)

As the owner of the RFC, I believe it's within your remit to do that work exploring and understanding the different ways we could solve this problem. Then you can present back your findings to us through the RFC, along with your recommendation and the reasons why that is preferable to the other approaches.

Part of exploring options might mean doing a quick implementation of a few different implementations until you understand well the pros and cons and can represent them fairly. We'll all pick up biases along the way when doing this sort of work, but as long as you logically and fairly represent them in the RFC, then that's ok and we can all make a fair and honest assessment.

If you don't have a strong recommendation, then please let me know if you're looking for some other outcome from this RFC as I might be misunderstanding what you're aiming to achieve :)

**Reply by sc15zs** [reply]
Thanks Sam. I don't have a strong recommendation, I'd like to achieve these two things:
1.  Do we think it's worth doing it, or not?   
2. If yes, then suggestions what people think would be suitable for our needs, if someone has experience with such a thing (rather than me trying some random SQLAlchemy library I come across and build something, and then being told that there are much better options).

If the consensus for the first point is that we want to have something in place, I'd be happy to get recommendations from you guys which 3rd party solution would be interesting to try, or what other custom solution would be possible I haven't thought of.

**Reply by samuelhwilliams** [reply]
Sorry I missed replying to this.

On point 1, I think it's something we should (or need to) do, so I'm favour of the concept.

On point 2, maybe we get together and have an hour talking through what it is that our requirements are for preserving/recording/auditing changes to DB models? I think if we put our heads together around a mural with a few other devs, maybe we can draw out clearly what some of those key needs are. With that it should then be easier to assess what options are available and think about implementation. Does that help? Happy to put some time in the calendar (might be in the New Year now, depending on yours and others time off?)

**Reply by sc15zs** [reply]
Thanks, Sam, but as I see it, we should postpone doing anything with this at the moment until there's a decision about the future of Submit and Find. 
It makes no sense to do any work, if senior leadership might decide to abandon them.

**Reply by samuelhwilliams** [reply]
Yep, that makes sense! Shall we close this discussion down for now then until we know more?

**Comment by sc15zs** [comment]
Closing until SLT decides about the future of Submit & Find.
